test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua

```diff
@@ -0,0 +1,69 @@
+local checktrace = require('jit.util').traceinfo
+local t = require('luatest')
+local g = t.group()
+
+-- Test payload to be recorded. It calculates the arithmetic sum
+-- of elements from 1 to the given n.
+local function sumN(n)
+    local result = 0
+    for i = 1, n do
+        result = result + i
+    end
+    return result
+end
+
+g.test_luajit_enable_checkhook = function()
+    -- Setup JIT engine to record a trace as soon as possible.
+    jit.opt.start('hotloop=1')
+    -- Purge all existing traces.
+    jit.flush()
+
+    -- Check there is no traces at the beginning of the test.
+    assert(not checktrace(1), "JIT flush doesn't work")
+    -- Run payload with the minimal possible number to force
+    -- the inner loop to be compiled.
+    sumN(3)
+    -- Check the new trace is compiled as a result of the payload
+    -- call above.
+    assert(checktrace(1), 'The loop in <sumN> is compiled')
+
+    -- Return JIT engine setup to its defaults.
+    jit.opt.start('hotloop=56')
+
+    -- Upvalues to fetch key and value from lines to be used in
+    -- the scope of the debug hook below.
+    local lidx, lval
+    -- Table with lines to be hit by debug hook below.
+    local lines = {
+        60, -- local sum = sumN(3)
+        8,  -- local result = 0
+        9,  -- for i = 1, n do
+        10, --   result = result + i -- (i is 1)
+        9,  -- for i = 1, n do
+        10, --   result = result + i -- (i is 2)
+        9,  -- for i = 1, n do
+        10, --   result = result + i -- (i is 3)
+        9,  -- for i = 1, n do
+        12, -- return result
+        63, -- debug.sethook()
+    }
+
+    -- Set debug hook.
+    debug.sethook(function(hookname, line)
+        -- XXX: Only line hooks are fired.
+        assert(hookname == 'line', 'Unexpected hook type is fired')
+        lidx, lval = next(lines, lidx)
```

Comment:
I'm really against such `next` usage here.

From the [Lua 5.1 Reference Manual](https://www.lua.org/manual/5.1/manual.html#pdf-next):
> The order in which the indices are enumerated is not specified, even for numeric indices. (To traverse a table in numeric order, use a numerical `for` or the [`ipairs`](https://www.lua.org/manual/5.1/manual.html#pdf-ipairs) function.) 

So, this test is fragile regarding internal `next` implementation. 
I suggest to use just numeric index incremented on each hook call to fetch line value from the table.

Reply:
Be-be-be, fine I'll rewrite it via numeric upvalue.